File: forgepay/packages/sdk-python/forgepay/resources/payments.py
from __future__ import annotations

from typing import Any

from forgepay._http import SyncTransport, AsyncTransport
from forgepay.types import Payment, PaymentCreateParams, Refund, ListResponse


class PaymentsResource:
    def __init__(self, transport: SyncTransport) -> None:
        self._t = transport

    def create(self, **kwargs: Any) -> Payment:
        params = PaymentCreateParams(**kwargs)
        body   = params.model_dump(exclude_none=True, exclude={"idempotency_key"})
        raw    = self._t.request(
            "POST", "/v1/payments",
            json=body,
            idempotency_key=params.idempotency_key,
        )
        return Payment.model_validate(raw)

    def retrieve(self, payment_id: str) -> Payment:
        return Payment.model_validate(self._t.request("GET", f"/v1/payments/{payment_id}"))

    def list(self, **params: Any) -> ListResponse[Payment]:  # type: ignore[type-arg]
        raw = self._t.request("GET", "/v1/payments", params=params or None)
        return ListResponse[Payment].model_validate(raw)

    def confirm(self, payment_id: str, **kwargs: Any) -> Payment:
        raw = self._t.request("POST", f"/v1/payments/{payment_id}/confirm", json=kwargs or None)
        return Payment.model_validate(raw)

    def cancel(self, payment_id: str) -> Payment:
        raw = self._t.request("POST", f"/v1/payments/{payment_id}/cancel")
        return Payment.model_validate(raw)

    def refund(self, payment_id: str, *, amount: int | None = None) -> Refund:
        body: dict[str, Any] = {"payment_id": payment_id}
        if amount is not None:
            body["amount"] = amount
        raw = self._t.request("POST", "/v1/refunds", json=body)
        return Refund.model_validate(raw)


class AsyncPaymentsResource:
    def __init__(self, transport: AsyncTransport) -> None:
        self._t = transport

    async def create(self, **kwargs: Any) -> Payment:
        params = PaymentCreateParams(**kwargs)
        body   = params.model_dump(exclude_none=True, exclude={"idempotency_key"})
        raw    = await self._t.request(
            "POST", "/v1/payments",
            json=body,
            idempotency_key=params.idempotency_key,
        )
        return Payment.model_validate(raw)

    async def retrieve(self, payment_id: str) -> Payment:
        return Payment.model_validate(await self._t.request("GET", f"/v1/payments/{payment_id}"))

    async def list(self, **params: Any) -> ListResponse[Payment]:  # type: ignore[type-arg]
        raw = await self._t.request("GET", "/v1/payments", params=params or None)
        return ListResponse[Payment].model_validate(raw)

    async def confirm(self, payment_id: str, **kwargs: Any) -> Payment:
        raw = await self._t.request(
            "POST", f"/v1/payments/{payment_id}/confirm", json=kwargs or None
        )
        return Payment.model_validate(raw)

    async def cancel(self, payment_id: str) -> Payment:
        raw = await self._t.request("POST", f"/v1/payments/{payment_id}/cancel")
        return Payment.model_validate(raw)

    async def refund(self, payment_id: str, *, amount: int | None = None) -> Refund:
        body: dict[str, Any] = {"payment_id": payment_id}
        if amount is not None:
            body["amount"] = amount
        raw = await self._t.request("POST", "/v1/refunds", json=body)
        return Refund.model_validate(raw)

File: forgepay/packages/sdk-js/src/resources/payments.ts
import type { FPHttpClient } from '../client.js';
import type {
  ListResponse,
  Payment,
  PaymentCreateParams,
  Refund,
  RefundCreateParams,
} from '../types.js';

export class PaymentsResource {
  constructor(private readonly client: FPHttpClient) {}

  /**
   * Create a payment intent.
   *
   * ```ts
   * const payment = await forgepay.payments.create({
   *   amount: 4900,
   *   currency: 'USD',
   *   customer_id: 'cus_123',
   *   idempotency_key: 'order_abc_001',
   * });
   * // → { id: 'pay_...', client_secret: '...', status: 'created' }
   * ```
   */
  async create(params: PaymentCreateParams): Promise<Payment> {
    const { idempotency_key, ...body } = params;
    return this.client.post<Payment>('/v1/payments', body, { idempotencyKey: idempotency_key });
  }

  /**
   * Retrieve a payment by ID.
   */
  async retrieve(paymentId: string): Promise<Payment> {
    return this.client.get<Payment>(`/v1/payments/${paymentId}`);
  }

  /**
   * List payments with optional filters.
   *
   * ```ts
   * const { data } = await forgepay.payments.list({ limit: 20, status: 'succeeded' });
   * ```
   */
  async list(params?: {
    limit?:       number;
    cursor?:      string;
    customer_id?: string;
    status?:      string;
    created_gte?: number;
    created_lte?: number;
  }): Promise<ListResponse<Payment>> {
    return this.client.get<ListResponse<Payment>>('/v1/payments', params as Record<string, string | number | undefined>);
  }

  /**
   * Cancel a payment that has not yet been captured.
   */
  async cancel(paymentId: string, reason?: string): Promise<Payment> {
    return this.client.post<Payment>(`/v1/payments/${paymentId}/cancel`, { reason });
  }

  /**
   * Refund a succeeded payment (full or partial).
   *
   * ```ts
   * // Full refund
   * const refund = await forgepay.payments.refund({ payment_id: 'pay_123' });
   *
   * // Partial refund of $10.00
   * const refund = await forgepay.payments.refund({ payment_id: 'pay_123', amount: 1000 });
   * ```
   */
  async refund(params: RefundCreateParams): Promise<Refund> {
    const { payment_id, ...body } = params;
    return this.client.post<Refund>(`/v1/payments/${payment_id}/refund`, body);
  }
}
